cluster-autoscaler/cloudprovider/gce/gce_manager.go

```diff
@@ -1048,6 +1075,45 @@ func (m *gceManagerImpl) GetResourceLimiter() (*cloudprovider.ResourceLimiter, e
 	return m.resourceLimiter, nil
 }
 
+func (m *gceManagerImpl) clearMachinesCache() error {
+	if m.machinesCacheLastRefresh.Add(machinesRefreshInterval).After(time.Now()) {
+		return nil
+	}
+	m.cacheMutex.Lock()
+	defer m.cacheMutex.Unlock()
+	m.machinesCache = make(map[machineTypeKey]*gce.MachineType)
+	m.machinesCacheLastRefresh = time.Now()
+	glog.V(2).Infof("Cleared machine types cache, next clear after %v", m.machinesCacheLastRefresh.Add(machinesRefreshInterval))
+	return nil
+}
+
+func (m *gceManagerImpl) fetchMachinesCache() error {
+	if m.machinesCacheLastRefresh.Add(machinesRefreshInterval).After(time.Now()) {
+		return nil
+	}
+	cluster, err := m.gkeAlphaService.Projects.Zones.Clusters.Get(m.projectId, m.location, m.clusterName).Do()
```

Comment:
I'm pretty sure this will segfault (nil ptr dereference) in ModeGKE. The reason is we only use gkeAlphaService in ModeGKENAP. In ModeGKE we use gkeService and / or gkeBetaService if on regional cluster. It's a horrible, horrible mess, but that's where we are.

Bottom line is you need to distinguish and use gkeAlphaService or gkeService depending on mode (I think you can still use gkeService in regional cluster - can you confirm @bskiba?)

Reply:
Not sure if Locations are already available in v1? If no you may actually need 3 implementations :(

Reply:
Locations is only available in beta now unfortunately. Generally, if we want to be sure that this works for regional clusters for the long run, we should use Location API instead of the Zone API so unfortunately gkeBetaService for regional :(

Reply:
Right, so once again we need 3 implementations of the same :( Let's try to clean this mess for 1.12.

Reply:
Actually, it looks like the docs are out of date, https://container.googleapis.com/$discovery/rest?version=v1 shows that the location API has been moved to v1. This means we should be able to eliminate gkeBetaService if we move all calls to the location path of the v1 api \o/

Reply:
Let's remove gkeBetaService in a separate PR though. This one is large enough and removing beta service will also be a big change. Good news is you will only need to have 2 versions in your new code (v1 service is also initialized in regional, so you can just use it in ModeGKE).

Reply:
done